Refactor MP API::unsubscribeFromLists to Doctrine [MAILPOET-4291]

costasovo opened the pull request:
[MAILPOET-4291]

[MAILPOET-4291]: https://mailpoet.atlassian.net/browse/MAILPOET-4291?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

costasovo committed:
Refactor getting subscriber and segments into a private methods

[MAILPOET-4291]

costasovo committed:
Refactor MP API unsubscribe from lists to doctrine

[MAILPOET-4291]

costasovo committed:
Move MP API unsubscribe from list test to a separate file

[MAILPOET-4291]

johnolek commented on mailpoet/lib/API/MP/v1/Subscribers.php:
```diff
@@ -84,61 +87,8 @@ public function subscribeToLists(
-      );
-      throw new APIException(sprintf($exception, implode(', ', $missingIds)), APIException::LIST_NOT_EXISTS);
-    }
+    $subscriber = $this->findSubscriber($subscriberId);
+    $foundSegments = $this->getAndValidateSegments($listIds, self::CONTEXT_SUBSCRIBE);
```
Do you think it might make sense to retrieve the segments before the subscriber just to be consistent with the old code? I doubt it really matters, but if someone made an API call with both an empty subscriber and no segment IDs, the previous code would return an error about requiring a segment while this would return an error about requiring a subscriber, right?

costasovo replied:
I was thinking that it was not a big deal. We can't fix it simply by changing the order of those methods because, in that case, we would change the order of subsequent checks. So I guess I could extract the basic checks of inputs and call them before so that we have consistent order of checks. 
I can imagine some plugin might be using an error message for its logic, so to be safe, I'll rather fix it.

johnolek commented on mailpoet/lib/API/MP/v1/Subscribers.php:
```diff
@@ -223,4 +186,92 @@ protected function _sendConfirmationEmail(SubscriberEntity $subscriberEntity) {
+   * @return SegmentEntity[]
+   * @throws APIException
+   */
+  private function getAndValidateSegments(array $listIds, string $context = self::CONTEXT_SUBSCRIBE): array {
```
Do you think it might be good to remove the default value for the `$context`? It looks like you're explicitly including the value everywhere this method is invoked, which I prefer to relying on default values. No big deal either way though.

costasovo replied:
Good point. I'll remove it.

johnolek reviewed (changes requested):
@costasovo This looks good to me overall! I'm only marking this as requesting changes because of one very small question about a difference in the order we're retrieving data and how that could result in different error messages before/after the refactor in one edge case.

The ticket also mentions checking/refactoring the `\MailPoet\API\MP\v1\CustomFields` class, but it looks like that class was already refactored (b368a7659375532b6e044161dff4ba4ac2f4e2e0). I just wanted to triple check with you that there's nothing else we need to do with the `CustomFields` class to consider all the requirements of the ticket complete.

costasovo committed:
Change order of MP API subscribe/unsubscribe calls checks to be backward compatible

In the previous version we checked the lists parameter before the subscriber.
[MAILPOET-4291]

costasovo committed:
Make a parameter in private method in MP API non-optional

[MAILPOET-4291]

costasovo commented:
@johnolek I've addressed your comments. Please see the two new commits.

> The ticket also mentions checking/refactoring the \MailPoet\API\MP\v1\CustomFields class, but it looks like that class was already refactored (https://github.com/mailpoet/mailpoet/commit/b368a7659375532b6e044161dff4ba4ac2f4e2e0). I just wanted to triple check with you that there's nothing else we need to do with the CustomFields class to consider all the requirements of the ticket complete.

I understood this mention of CustomFields as that I should get inspired by how the refactor was done for the custom field. It is mentioned in all MP API related tickets.

johnolek reviewed (approved):
Thanks for making those changes! And sorry about misinterpreting the comment in the ticket about the other code!
